src/rust/engine/process_execution/src/remote.rs

```diff
@@ -152,33 +161,59 @@ impl CommandRunner {
                             // Reset `iter_num` on `MissingDigests`
                             .map(|operation| future::Loop::Continue((operation, 0)))
                             .to_boxed()
-                      },
-                      ExecutionError::NotFinished(operation_name) => {
-                        let mut operation_request =
-                          bazel_protos::operations::GetOperationRequest::new();
-                        operation_request.set_name(operation_name.clone());
-
-                        let backoff_period = min(
-                      CommandRunner::BACKOFF_MAX_WAIT_MILLIS,
-                      (1 + iter_num) * CommandRunner::BACKOFF_INCR_WAIT_MILLIS);
-
-                        let grpc_result = map_grpc_result(
-                    operations_client.get().get_operation(&operation_request));
-
-                        Delay::new(Duration::from_millis(backoff_period))
-                          .map_err(move |e| format!(
-                            "Future-Delay errored at operation result polling for {}: {}",
-                              operation_name, e))
-                            .and_then(move |_| {
-                              future::ok(
-                             future::Loop::Continue(
-                                  (try_future!(grpc_result), iter_num + 1))).to_boxed()
-                            }).to_boxed()
-                      },
                     }
-                  })
-              })
-            }).to_boxed()
+                    ExecutionError::NotFinished(operation_name) => {
+                      let mut operation_request =
+                        bazel_protos::operations::GetOperationRequest::new();
+                      operation_request.set_name(operation_name.clone());
+
+                      let backoff_period = min(
+                        CommandRunner::BACKOFF_MAX_WAIT_MILLIS,
+                        (1 + iter_num) * CommandRunner::BACKOFF_INCR_WAIT_MILLIS,
+                      );
+
+                      let grpc_result =
+                        map_grpc_result(operations_client.get().get_operation(&operation_request));
+                      // take the grpc result and cancel the op if too much time has passed.
+                      match grpc_result {
+                        Ok(operation) => {
+                          match start_time.elapsed() {
+                            Ok(elapsed) => {
+                              if elapsed > timeout {
+                                future::err(format!(
+                                  "Exceeded time out of {:?} with {:?} for operation {}, {}",
+                                  timeout, elapsed, operation_name, req_description
+                                )).to_boxed() as BoxFuture<_, _>
```

Comment:
I don't think the `as BoxFuture<_, _>` is needed any more, since https://github.com/pantsbuild/pants/pull/5754 (and throughout)

Reply:
nice